Lib/fontTools/designspaceLib/__init__.py

```diff
@@ -735,6 +735,21 @@ def read(self):
         self.readInstances()
         self.readLib()
 
+    def validateAttributes(self, element, attributeTypeMap):
+        for attributeName, attributeType in attributeTypeMap.items():
+            attrib = element.attrib.get(attributeName)
+            if attrib is None:
+                raise DesignSpaceDocumentError(
+                    "<%s> element missing required attribute '%s'"
```

Comment:
the required-ness of attributes is a pain point in designspace spec and lib (I commented on that somewhere, forgot where exactly). I believe the code was written deliberately to allow most of the attributes to be optional (you can tell by the many `.get(something)` that don't handle `None`-ness). So basically clients are supposed to check for that and handle missing data.
For example, here you made both axis.name and axis.tag required; I think varLib works even if one of the two is missing (attempting to derive one from the other).
Before we enforce required-ness with an exception like you did here, we should review thoroughly all the piece of data that we think should be required, make that explicit in the spec and adjust the code accordingly (while avoiding possible breakages in other tools).

Reply:
An XML schema for designspace would be super helpful...

Reply:
I don't know how entangled designspace and UFO are (the use of `lib` and `info` suggest that they are, and I was quite surprised when I realised that all the filenames could only refer to .ufo files, not .ttfs), but if so maybe that's something to bring up at Friday's UFO meeting.

Reply:
with varLib as consumer, source filenames can also be .ttf or even .ttx files.

> maybe that's something to bring up at Friday's UFO meeting

there's a lot on the table already but sure, you can try